Issue#804 Add Observable.toInputStream

melgenek opened the pull request:
This change introduces a simple `toInputStream` method #804 

The simplifications:
1) The `read` on InputSteam is blocking without prefetch. So `AsyncVar` is used instead of a queue.
- fs has blocking queue with 1 element
[blocking call](https://github.com/functional-streams-for-scala/fs2/blob/series/1.0/io/src/main/scala/fs2/io/JavaInputOutputStream.scala#L105) ; [queue factory method](https://github.com/functional-streams-for-scala/fs2/blob/series/1.0/core/shared/src/main/scala/fs2/concurrent/Queue.scala#L173-L175)
- akka-streams has a configurable buffer size 
[queue creation](https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala#L53)
2) The blocking to wait for next element is infinite
- fs2 seems to block as well for infinite time
- akka has [a read timeout](https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala#L42)
3) this implementation is based on Task. I could also add [Cats Resource](https://typelevel.org/cats-effect/api/cats/effect/Resource.html) if the implementation would be considered sufficient and correct.

Avasil commented:
Hello @melgenek thank you for the PR!
I will try to review it in the next couple of days. I'm sorry for the long delay, I've been sick lately and limited my open source time to minimum

melgenek commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamResource.scala:
```diff
@@ -0,0 +1,128 @@
+  private type Bytes = Array[Byte]
+
+  def toTask: Task[InputStream] = {
+    Task.create[InputStream] { (s, cb) =>
```
Well, yes. But it does not have to live in the monix-reactive, because resource comes from cats

Avasil replied:
We depend on Cats-Effect since `3.x` line so we have it available

melgenek replied:
introduced the `Resource`

melgenek commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamResource.scala:
```diff
@@ -0,0 +1,128 @@
+      }
+
+      def onComplete(): Unit = {
+        state.put(None).value
```
I had a wrong guess, that this blocks until the `put` succeeds. Yes, `.value` is useless here

melgenek commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamResource.scala:
```diff
@@ -0,0 +1,128 @@
+      }
+
+      def onError(ex: Throwable): Unit = {
+        state.put(None).value
```
true

melgenek commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamResource.scala:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    override def read(arr: Array[Byte], begin: Int, length: Int): Int = {
+      require(arr.length > 0, "array size must be >= 0")
```
This is something that I saw in the akka-streams code. Looks pretty logical, as this is sanity validation.
Instead of this validation I could:
- do nothing if array is empty
- do nothing if begin/length are negative 
- read only the amount of bytes that fit into the buffer.

However, as a user of the method I would prefer seeing exception, then having strange results in the byte array, that are caused by some internal implementation details.

Avasil replied:
Makes sense!

melgenek commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamResource.scala:
```diff
@@ -0,0 +1,128 @@
+  private def subscribe(s: Scheduler): (AsyncVar[Option[Bytes]], Cancelable) = {
+    val state = AsyncVar.empty[Option[Bytes]]()
+
+    val cancelable = observable.executeAsync.unsafeSubscribeFn(new Subscriber.Sync[Array[Byte]] {
```
it is not needed. I had a wrong assumption that `.value` blocks. `state.put(Some(elem)).map(_ => Continue)` is sufficient for this case

melgenek commented on monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamResource.scala:
```diff
@@ -0,0 +1,128 @@
+      implicit val scheduler = s
+
+      override def onNext(elem: Array[Byte]): Ack = {
+        state.put(Some(elem)).value
```
it looks like `state.put(Some(elem)).map(_ => Continue)` causes `onComplete` to be delayed even if the stream (and consequently Observable) is cancelled. I will try to figure this out.

Avasil replied:
Ah, right! 
Let me know if you get stuck

melgenek replied:
@Avasil I am really a little stuck.There is a test case:
```
  testAsync("InputStream waits until enough bytes are available") {
    val array1 = randomByteArrayOfSize(10)
    val array2 = randomByteArrayOfSize(5)
    val array3 = randomByteArrayOfSize(5)
    var completed = false
    val observable = Observable.fromIterable(Seq(array1, array2, array3))
      .doOnComplete(Task {
        completed = true
      })

    Observable.toInputStream(observable)
      .map { inputStream =>
        val result = new Array[Byte](20)
        inputStream.read(result)

        assertEquals(result.toList, (array1 ++ array2 ++ array3).toList)
        assertEquals(inputStream.read(), -1.toByte)
        assert(completed)
      }
      .runToFuture(monix.execution.Scheduler.global)
  }
```
- I create a stream of 3 arrays
- I am trying to read 20 bytes
- and 2 out of 3 arrays come to `onNext`
- and both 2 are read in `ensureBufferSize`
- but then there are no `onNext` calls and the test hangs

This is visible in the ci test run https://travis-ci.org/monix/monix/builds/609307795?utm_source=github_status&utm_medium=notification

I tryied checking threads and they seem to be different in the `onNext`/`ensureBufferSize` so the execution has to continue, as I understand.
I also tried moving to built-in operators, because it looks like SafeSubscriber manages cancellations properly and I hoped that other operators at least handle concurrency properly. But it does not help either.

Could you please suggest how to debug this, because I do not where the next `onNext` call is stuck? 
Thank you.

Avasil replied:
I don't know why the test has failed, it passed on CI after restart. I will try few more times to see if it comes back.

Does it also happen locally?

melgenek replied:
yes, this happens locally

Avasil replied:
How often does it happen? And do you run only this single test or all of them?
I'm trying to reproduce but so far it's passing on my machine

melgenek replied:
Well, I can observe this once in several runs locally. 
I run it like this `sbt "testOnly monix.reactive.internal.builders.ToInputStreamSuite"`
There is the thread dump from the last run that stuck https://gist.github.com/melgenek/11ab38be0b1098cd732960f18a7ef92b

Avasil replied:
I see, I will try again tomorrow and investigate

Avasil replied:
I tried on different computer and I can finally reproduce it! I'll let you know when I figure it out

Avasil replied:
BTW does it happen if we use `Scheduler.io` instead of `global`  or wrap `Await.result` in `blocking { }` ? 
If I add a timeout to `Await.result` then it seems like the next `onNext` continues after `inputStream.read()` times out and in one case I reproduced it happened on the same thread for some reason. Maybe it was the only thread available but that seems weird

melgenek replied:
1) `blocking` did not help
2) I cannot reproduce the problem with `io` pool any more

melgenek commented:
@Avasil I updated the code according to your previous comments: added Resource, used the Async Subscriber, added the configurable timeout for `read`. Additionally, the exceptions are now propagated from Observable into the `read` call and wrapped into IOException (it is the one, that is expected to be thrown in `read` method).

The only thing left is to move the code somewhere outside the `builders`. But you have to say where.

Avasil commented:
@melgenek 
Thank you, I will re-review after the weekend and play with it a little bit to search for corner cases

> The only thing left is to move the code somewhere outside the builders. But you have to say where.

I will try to reach @alexandru to get his opinion.
Adding `monix-io` module would be a nice option but if it's here then it probably needs to have a different versioning
Although `monix-reactive` could be fine as well to have a symmetry with `fromInputStream`
